app/services/course/assessment/answer/text_response_auto_grading_service.rb

```diff
@@ -77,13 +84,271 @@ def explanations_for(solutions)
     solutions.map(&:explanation).tap(&:compact!)
   end
 
-  # Mark the correctness of the answer based on solutions
+  # Mark the correctness of the answer based on solutions.
   #
   # @param [Course::Assessment::Question::TextResponse] question The question answered by the
   #   student.
   # @param [Array<Course::Assessment::Question::TextResponseSolution>] solutions The solutions that
   #   matches the student's answer.
+  # @param [Boolean] correct True if the answer is correct.
   def correctness_for(question, solutions)
     solutions.map(&:grade).sum >= question.maximum_grade
   end
+
+  # Grades the given answer for comprehension questions.
+  #
+  # @param [Course::Assessment::Answer::TextResponse] answer The answer specified by the
+  #   student.
+  # @return [Array<(Boolean, Integer, Object)>] The correct status, grade and the messages to be
+  #   assigned to the grading.
+  def evaluate_answer_comprehension(answer)
+    question = answer.question.actable
+    answer_text = answer.normalized_answer_text
+    answer_text_array = answer_text.downcase.gsub(/([^a-z ])/, ' ').split(' ')
+    answer_text_lemma_array = []
+    answer_text_array.each { |a| answer_text_lemma_array.push(WordNet::Synset.morphy_all(a).first || a) }
+
+    answer_text_lemma_status = { 'compre_lifted_word' => Array.new(answer_text_lemma_array.length, nil),
```

Comment:
This is really C style. Could you see if you could do away with passing references around and mutating arrays?

Perhaps move this below and try to do something like:

```ruby
answer_text_lemma_status = {
  compre_lifted_word: find_compre_lifted_word_in_answer(...),
  compre_keyword: find_compre_keyword_in_answer(...),
}
```